### ISSUE TITLE
Proxy Protocol on health check port

### ISSUE THREAD

**Issue by jgournet** [istio/istio#39868]
**Bug description**
Setup:
- NLB with TLS termination, and proxy protocol enabled
- proxy protocol enabled on istio with:
```
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: proxy-protocol
  namespace: istio-system
spec:
  configPatches:
  - applyTo: LISTENER
    patch:
      operation: MERGE
      value:
        listener_filters:
        - name: envoy.listener.proxy_protocol
        - name: envoy.listener.tls_inspector
  workloadSelector:
    labels:
      istio: ingressgateway
```
- Target Group health check with HTTP test pointing to /healthz/ready, on the health check port.

Result: the AWS instances are marked as "unhealthy": the istio-ingress gateway replies with 400.

Note: if I disable proxy protocol on the target group, the instances are marked as healthy after ~30seconds.

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Expected behavior**
istio should be able to handle proxy protocol on its healthcheck port.

**Steps to reproduce the bug**
See the setup above.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
```
istioctl version --remote
client version: 1.14.0
control plane version: 1.14.0
data plane version: 1.14.0 (269 proxies)
```
**How was Istio installed?**
via helm chart

**Environment where bug was observed (cloud vendor, OS, etc)**
AWS

Note:
This is a reopening of ticket:
https://github.com/istio/istio/issues/25875

**Comment by hzxuzhonghu** [comment]
According to envoy doc https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/listener_filters/proxy_protocol#proxy-protocol

>Note: if the filter is enabled, the Proxy Protocol must be present on the connection (either version 1 or version 2), the standard does not allow parsing to determine if it is present or not.

**Comment by hzxuzhonghu** [comment]
Try setting allow  allow_requests_without_proxy_protocol https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/listener/proxy_protocol/v3/proxy_protocol.proto#extensions-filters-listener-proxy-protocol-v3-proxyprotocol

**Comment by jgournet** [comment]
> Try setting allow allow_requests_without_proxy_protocol https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/listener/proxy_protocol/v3/proxy_protocol.proto#extensions-filters-listener-proxy-protocol-v3-proxyprotocol

Thanks ! that sounds indeed promising !
Sorry for asking, but would you have the code to change this via istio ? I guess I could find it out by myself, but that will take me a lot of trial and error

**Comment by hzxuzhonghu** [comment]
Sorry, i don't have that config

**Comment by paul-conneqtech** [comment]
@jgournet In the previous ticket you mentioned a workaround for this, could you share the EnvoyFilter you used for this?

**Comment by jgournet** [comment]
@paul-conneqtech : it's actually not an EnvoyFilter :) 
Here's the config, with comments included to help:
```
---
# Background:
# NLB health checks are sent with proxy protocol v2, since the target group is set that way.
# BUT: istio status port (32020 -> 15020) does not expect proxy protocol, so the calls fail.
# => Workaround: 
# we set the target group to use the traffic port for the health check
# and we add this Gateway+Virtual Service to forward to the health check port
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: health-check
  namespace: istio-system
spec:
  selector:
    istio: ingressgateway
  servers:
  - hosts:
    - '*'
    port:
      name: http
      number: 80
      protocol: HTTP2
  - hosts:
    - '*'
    port:
      name: https
      number: 443
      protocol: HTTP2
---
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: health-check
  namespace: istio-system
spec:
  gateways:
  - istio-system/health-check
  hosts:
  - "*"
  http:
  - name: "health-check"
    match:
    - uri:
        prefix: "/healthz/ready"
    route:
    - destination:
        host: istio-ingressgateway
        port:
          number: 15020
```

Note: I'm still trying to find time to try out " allow_requests_without_proxy_protocol" from above comments ...

**Comment by AngeloDanducci** [comment]
I have run into  the same/similar issue, ie istio does not seem to handle proxy protocol on its healthcheck port. 

Adding comment here to bring visibility to the issue 👍

**Comment by yann-soubeyrand** [comment]
What about having a second health check port with proxy protocol enabled? We do this by adding an HTTP listener on port 15022 on our gateway and by using the following Envoy filter:

```
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: proxy-protocol
spec:
  configPatches:
    - applyTo: NETWORK_FILTER
      match:
        context: GATEWAY
        listener:
          portNumber: 15022
          filterChain:
            filter:
              name: envoy.filters.network.http_connection_manager
      patch:
        operation: REPLACE
        value:
          name: envoy.filters.network.http_connection_manager
          typed_config:
            '@type': type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
            stat_prefix: agent
            route_config:
              virtual_hosts:
                - name: backend
                  domains:
                    - '*'
                  routes:
                    - match:
                        prefix: /healthz/ready
                      route:
                        cluster: agent
            http_filters:
              - name: envoy.filters.http.router
                typed_config:
                  '@type': type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
    targetRefs:
      - group: gateway.networking.k8s.io
        kind: Gateway
        name: istio-gateway
```

Maybe this should be part of the bootstrap configuration alongside the configuration of the port 15021 (since we cannot modify the configuration of the latter)?

**Comment by BeyondEvil** [comment]
So what's the real fix here?

**Comment by yann-soubeyrand** [comment]
@BeyondEvil the real fix is to use this new v1.27 feature: https://istio.io/latest/docs/reference/commands/pilot-agent/#envoy-status-port-enable-proxy-protocol.

**Comment by jgournet** [comment]
THANK YOU @yann-soubeyrand for the info ! 
(too bad we're not using NLB anymore by now :D but it's good to see that kind of new feature)

**Comment by Mo0rBy** [comment]
> [@BeyondEvil](https://github.com/BeyondEvil) the real fix is to use this new v1.27 feature: https://istio.io/latest/docs/reference/commands/pilot-agent/#envoy-status-port-enable-proxy-protocol.

What component of Istio does this envVar need to be applied to?
I have each component installed via the official Helm charts:
1. istio base
2. istio-cni
3. istio-gateway
4. istiod

I have applied this envVar to istio-gateway as thats where proxy protocol is initially enabled and where the EnvoyFilter's are doing some modifications, so my guess was that was the component to add `ENVOY_STATUS_PORT_ENABLE_PROXY_PROTOCOL=true` to, but I've done that and I still get "Unhealthy" statuses on my instance targets.

**Comment by Mo0rBy** [comment]
> [@paul-conneqtech](https://github.com/paul-conneqtech) : it's actually not an EnvoyFilter :) Here's the config, with comments included to help:
> 
> ```
> ---
> # Background:
> # NLB health checks are sent with proxy protocol v2, since the target group is set that way.
> # BUT: istio status port (32020 -> 15020) does not expect proxy protocol, so the calls fail.
> # => Workaround: 
> # we set the target group to use the traffic port for the health check
> # and we add this Gateway+Virtual Service to forward to the health check port
> apiVersion: networking.istio.io/v1beta1
> kind: Gateway
> metadata:
>   name: health-check
>   namespace: istio-system
> spec:
>   selector:
>     istio: ingressgateway
>   servers:
>   - hosts:
>     - '*'
>     port:
>       name: http
>       number: 80
>       protocol: HTTP2
>   - hosts:
>     - '*'
>     port:
>       name: https
>       number: 443
>       protocol: HTTP2
> ---
> apiVersion: networking.istio.io/v1beta1
> kind: VirtualService
> metadata:
>   name: health-check
>   namespace: istio-system
> spec:
>   gateways:
>   - istio-system/health-check
>   hosts:
>   - "*"
>   http:
>   - name: "health-check"
>     match:
>     - uri:
>         prefix: "/healthz/ready"
>     route:
>     - destination:
>         host: istio-ingressgateway
>         port:
>           number: 15020
> ```
> 
> Note: I'm still trying to find time to try out " allow_requests_without_proxy_protocol" from above comments ...

This _almost_ works for me.
The only issue is that I have a requirement to use port 22 as my cluster contains a GitLab cluster that needs port 22 for Git operations and using the traffic port for the target group using port 22 breaks this.

**Comment by Mo0rBy** [comment]
I tried changing it to this, but still doesn't work:
```
---
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: proxy-protocol-fix-health-check
spec:
  selector:
    istio: gateway
  servers:
    - hosts:
        - '*'
      port:
        name: istio-status
        number: 15021
        protocol: TCP
---
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: proxy-protocol-fix-health-check
spec:
  gateways:
    - istio-system/proxy-protocol-fix-health-check
  hosts:
    - "*"
  http:
    - name: "health-check"
      match:
        - uri:
            prefix: "/healthz/ready"
      route:
        - destination:
            host: istio-gateway
            port:
              number: 15021
```

and my istio ingress gateway annotations:
```
      annotations:
        service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
        service.beta.kubernetes.io/aws-load-balancer-internal: "true"
        service.beta.kubernetes.io/aws-load-balancer-name: "dso"
        service.beta.kubernetes.io/aws-load-balancer-security-groups: "sg-04a585782f6ce0359"
        service.beta.kubernetes.io/aws-load-balancer-target-group-attributes: "preserve_client_ip.enabled=false"
        service.beta.kubernetes.io/aws-load-balancer-proxy-protocol: "*"
        service.beta.kubernetes.io/aws-load-balancer-healthcheck-protocol: "TCP"
        service.beta.kubernetes.io/aws-load-balancer-healthcheck-port: "15021"
        service.beta.kubernetes.io/aws-load-balancer-healthcheck-path: "/healthz/ready"

**Comment by Mo0rBy** [comment]
I've even tried adding the `allow_requests_without_proxy_protocol` to `true` on my EnvoyFilter and still get failing healthchecks:
```
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: enable-proxy-protocol
spec:
  workloadSelector:
    labels:
      istio: gateway
  configPatches:
    - applyTo: LISTENER_FILTER
      match:
        listener:
          listenerFilter: "envoy.filters.listener.tls_inspector"
      patch:
        operation: INSERT_BEFORE
        value:
          name: envoy.filters.listener.proxy_protocol
          typed_config:
            '@type': type.googleapis.com/envoy.extensions.filters.listener.proxy_protocol.v3.ProxyProtocol
            "allow_requests_without_proxy_protocol": true
```

**Comment by Kyslik** [comment]
Regarding proxy protocol also see network topologies (alpha) https://istio.io/latest/docs/ops/configuration/traffic-management/network-topologies/#proxy-protocol.

**Comment by Mo0rBy** [comment]
I figured it out this weekend, I needed to use the NodePort that maps to the istio status port for my load balancer health check (30567 or whatever gets assigned when the LoadBalancer service is created), and the rest is pretty much the same.

I have a new issue now of port 22 not working with proxy protocol for things like Git over SSH, and I think the only solution is a separate LoadBalancer for port 22.
I think this ok for my purposes though.
